### ISSUE TITLE
Number of connections between client and server

### ISSUE THREAD

**Issue by hi-Mr-ye** [sysrepo/sysrepo#2688]
Hi @michalvasko ,
I create a session, and server printing [INF]: SR: Session 3422 (user "administrator", CID 3) created.
then I create another session and server printing [INF]: SR: Session 3423 (user "administrator", CID 3) created.
next the server printing [ERR]: SR: Callback event "notif" with ID 220 processing timed out.
After that, the session is disconnected. Is it the result of too many connected sessions?

The following is all the information printed by the server:
[INF]: LN: Accepted a connection on 0.0.0.0:830 from 192.168.1.198:52464.
[INF]: SR: Session 3422 (user "administrator", CID 3) created.
[INF]: SR: Published event "notif" "ietf-netconf-notifications" with ID 220 priority 0 for 2 subscribers.
[INF]: LN: Received an SSH message "request-service" of subtype "ssh-userauth".
[INF]: LN: Received an SSH message "request-auth" of subtype "none".
[INF]: LN: Received an SSH message "request-auth" of subtype "password".
[INF]: LN: User "administrator" authenticated.
[INF]: LN: Received an SSH message "request-channel-open" of subtype "session".
[INF]: LN: Received an SSH message "request-channel" of subtype "subsystem".
[INF]: SR: Session 3423 (user "administrator", CID 3) created.
[ERR]: SR: Callback event "notif" with ID 220 processing timed out.
[INF]: NP: Generated new event (netconf-session-end).
[INF]: SR: Published event "notif" "ietf-netconf-notifications" with ID 221 priority 0 for 2 subscribers.
[ERR]: LN: Session 138: SSH channel poll error (Received SSH_MSG_DISCONNECT: 11:Closed due to user request.).
[INF]: NP: Session 138: thread 1 event session terminated.
[ERR]: SR: Callback event "notif" with ID 221 processing timed out.
[INF]: NP: Generated new event (netconf-session-start).
[INF]: SR: Published event "notif" "ietf-netconf-notifications" with ID 222 priority 0 for 2 subscribers.
[INF]: LN: Session 140: Received an SSH message "request-channel-open" of subtype "session".
[INF]: SR: Published event "rpc" "/notifications:create-subscription" with ID 68 priority 0 for 1 subscribers.
[ERR]: SR: Waiting on a conditional variable failed (sr_shmext_notif_sub_add: Connection timed out).
[INF]: SR: Failed processing of "rpc" event with ID 67 priority 0 (after timeout or earlier error).
[INF]: SR: Processing "/notifications:create-subscription" "rpc" event with ID 68 priority 0 (remaining 1 subscribers).
[ERR]: SR: Callback event "notif" with ID 222 processing timed out.
[INF]: NP: Generated new event (netconf-session-end).
[INF]: SR: Published event "rpc" "/ietf-netconf:get" with ID 581 priority 0 for 1 subscribers.
[INF]: SR: Published event "notif" "ietf-netconf-notifications" with ID 223 priority 0 for 2 subscribers.
[ERR]: SR: Callback event "rpc" with ID 68 processing timed out.
[WRN]: SR: Event "rpc" with ID 68 priority 0 failed (Timeout expired).
[ERR]: SR: User callback failed.
[ERR]: NP: Failed to send an RPC (User callback failed).
[INF]: NP: Session 139: thread 2 event new RPC.
[INF]: NP: Session 139: thread 2 event reply error.
[ERR]: SR: Callback event "rpc" with ID 581 processing timed out.
[WRN]: SR: Event "rpc" with ID 581 priority 0 failed (Timeout expired).
[ERR]: SR: User callback failed.
[ERR]: NP: Failed to send an RPC (User callback failed).
[ERR]: SR: Callback event "notif" with ID 223 processing timed out.
[INF]: NP: Generated new event (netconf-session-start).

**Comment by michalvasko** [comment]
If the timeout occurs only after reaching 3k+ sessions, it is possible that is the cause of it. There are no optimizations concerning session lookup because it was not needed. But you really have more than 3k simultaneous sessions? I do not think there is any sensible reason for that so you should probably try to stop whatever sessions are no longer needed.

**Comment by hi-Mr-ye** [comment]
I use a client tool that takes a long time to connect to the server, so I constantly create sessions to connect to the server  
and client tool is 
<dependency>
			<groupId>com.airlenet.netconf</groupId>
			<artifactId>netconf-client</artifactId>
			<version>2.0.0</version>
		</dependency>
Is it a client tool problem?

**Comment by michalvasko** [comment]
You are mixing things. The issue you described hints that there are lots of *sysrepo* sessions, which may cause some problems although ideally it should not. Now you are talking about a NETCONF client, which creates *NETCONF* sessions, which are something different altogether because they are established between `netopeer2` and the NETCONF client, *sysrepo* sessions are between `netopeer2` and `sysrepo`.

But seeing that you are using a Java NETCONF client (is my guess based on the `groupId`) it may really be just a problem of short timeout, you can try increasing it by running `netopeer2-server` with parameter `-t 10`, for example.

**Comment by hi-Mr-ye** [comment]
oh I'm sorry! I'm confused about the two. You said earlierreaching 3k+ sessions, This session refers tobetween netopeer2 and the NETCONF client or sysrepo sessions are between netopeer2 and sysrepo?

**Comment by michalvasko** [comment]
> [INF]: SR: Session 3422 (user "administrator", CID 3) created.

The "prefix" of the message is SR, which is short for *sysrepo* so we are talking about *sysrepo* sessions. You can also see prefixes LN (libnetconf2) and NP (netopeer2), which also create sessions but with much smaller IDs. But please refer to what I wrote before, I am not sure the number of the sessions is really the problem although it is not out of the question.

**Comment by hi-Mr-ye** [comment]
I understand！thank you!
